presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotBrokerPageSourceBase.java

```diff
@@ -0,0 +1,384 @@
+                results[i] = asText(node.get(i));
+            }
+            return Arrays.toString(results);
+        }
+        return node.isNull() ? null : node.asText();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return 0;
+    }
+
+    @Override
+    public void close()
+    {
+        finished = true;
+    }
+
+    protected int issueQueryAndPopulate(
+            PinotQueryGenerator.GeneratedPinotQuery pinotQuery,
+            List<BlockBuilder> blockBuilders,
+            List<Type> types)
+    {
+        return doWithRetries(PinotSessionProperties.getPinotRetryCount(session), (retryNumber) -> {
+            String queryHost;
+            Optional<String> rpcService;
+            if (pinotConfig.getRestProxyUrl() != null) {
+                queryHost = pinotConfig.getRestProxyUrl();
+                rpcService = Optional.ofNullable(pinotConfig.getRestProxyServiceForQuery());
+            }
+            else {
+                queryHost = clusterInfoFetcher.getBrokerHost(pinotQuery.getTable());
+                rpcService = Optional.empty();
+            }
+            Request.Builder builder = Request.Builder
+                    .preparePost()
+                    .setUri(URI.create(String.format(getQueryUrlTemplate(), queryHost)));
+            String body = clusterInfoFetcher.doHttpActionWithHeaders(builder, Optional.of(String.format(getRequestPayloadTemplate(), pinotQuery.getSql())), rpcService);
+            return populateFromQueryResults(pinotQuery, blockBuilders, types, body);
+        });
+    }
+
+    abstract String getQueryUrlTemplate();
+
+    abstract String getRequestPayloadTemplate();
+
+    abstract int populateFromQueryResults(PinotQueryGenerator.GeneratedPinotQuery pinotQuery, List<BlockBuilder> blockBuilders, List<Type> types, String body);
+
+    public class BlockAndTypeBuilder
+    {
+        private final PageBuilder pageBuilder;
+        private final List<BlockBuilder> columnBlockBuilders;
+        private final List<Type> columnTypes;
+
+        public BlockAndTypeBuilder(PageBuilder pageBuilder, List<BlockBuilder> columnBlockBuilders, List<Type> columnTypes)
+        {
+            this.pageBuilder = pageBuilder;
+            this.columnBlockBuilders = columnBlockBuilders;
+            this.columnTypes = columnTypes;
+        }
+
+        public PageBuilder getPageBuilder()
+        {
+            return pageBuilder;
+        }
+
+        public List<BlockBuilder> getColumnBlockBuilders()
+        {
+            return columnBlockBuilders;
+        }
+
+        public List<Type> getColumnTypes()
+        {
+            return columnTypes;
+        }
+    }
+
+    @VisibleForTesting
+    public BlockAndTypeBuilder buildBlockAndTypeBuilder(List<PinotColumnHandle> columnHandles, PinotQueryGenerator.GeneratedPinotQuery brokerPql, List<PinotColumnHandle> expectedHandles)
+    {
+        // When we created the PQL, we came up with some column handles
```

Comment:
The whole notion of expected column handles comes from the fact that the PQL end point returns the group by results first before the aggregations. (and that it fails with not having aggregations and thus we have to introduce fake aggregations).

I am wondering if this constraint also holds for the SQL end point ? if not, then I think we should push this method to be abstract here.

Reply:
Let me double check on that.
From sql endpoint, we removed the constraint, so expectedColumnHandles should be same as columnHandles

Reply:
move `expectedColumnHandles` to `PinotBrokerPageSourcePql`